subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/ClassSetAnalysisUpdater.java

```diff
@@ -65,8 +72,51 @@ public void updateAnalysis(JavaCompileSpec spec) {
         for (File baseDir : baseDirs) {
             fileOperations.fileTree(baseDir).visit(analyzer);
         }
+        if (spec.getCompileOptions().isIncrementalAnnotationProcessing()) {
+            recordGeneratedTypeDependencies(analyzer, spec);
+        }
         ClassSetAnalysisData data = analyzer.getAnalysis();
         stash.put(data);
         LOG.info("Class dependency analysis for incremental compilation took {}.", clock.getElapsed());
     }
+
+    private void recordGeneratedTypeDependencies(ClassFilesAnalyzer analyzer, JavaCompileSpec spec) {
+        try {
+            File workingDir = spec.getIncrementalAnnotationProcessorWorkingDir();
+            // Create the incap working dir if necessary.  Incap will throw if the dir doesn't exist.
+            if (!workingDir.exists()) {
+                workingDir.mkdirs();
+            }
+            if (!workingDir.exists()) {
+                throw new IllegalStateException("Error creating: " + workingDir);
+            }
+            Map<InputType, Set<GeneratedFile>> mappings
+                = IncapBuildClientFactory.getBuildClient(workingDir).readMappingFile().getMapInputTypeToGeneratedFiles();
+            for (Map.Entry<InputType, Set<GeneratedFile>> entry : mappings.entrySet()) {
+                String generatingType = entry.getKey().getName();
+                String generatedType = null;
+                for (GeneratedFile output : entry.getValue()) {
+                    switch (output.getType()) {
+                        case CLASS:
+                            // TODO:  This needs to be tested.
+                            generatedType = ((GeneratedClassFile)output).getName().toString();
+                            break;
+                        case SOURCE:
+                            generatedType = ((GeneratedSourceFile)output).getName().toString();
+                            break;
+                        case RESOURCE:
+                            // For now, don't allow these.
+                            // TODO:  Incap should somehow report which AP it was, for logging.
+                            // Might need to build metadata into the mappings file.
+                            throw new IllegalStateException("Incremental annotation processor found that generates a resource.");
+                    }
+                    analyzer.addGeneratorMapping(generatingType, generatedType);
+                }
+            }
+        } catch (IOException iox) {
+            // We don't currently have a way to ask the Incap library whether the mappings file exists.
```

Comment:
I need some guidance on how to back out into a full build here if we encounter an error.  JavaCompile has already determined that the build is going to be incremental.  I seem to recall some code somewhere that changes the build type to non-incremental -- I think it might have been the old check for APs present.  Do you remember offhand?